docs/dev-guides/java-style.md

```diff
@@ -0,0 +1,166 @@
+
+Guidance for **unchecked exceptions**:
+
+*   More specific exceptions _should_ be thrown instead of generic RuntimeExceptions.
+    *   IllegalArgumentException, ConcurrentModificationException, UnsupportedOperationException are commonly reused exceptions from the Java library.
+    *   Other Terra-specific runtime exceptions _should_ be centralized in [terra-common-lib](https://github.com/DataBiosphere/terra-common-lib/tree/develop/src/main/java/bio/terra/common/exception).
+*   Service repos may also house application-specific exceptions ([WSM example](https://github.com/DataBiosphere/terra-workspace-manager/tree/dev/src/main/java/bio/terra/workspace/common/exception)).
+*   In many cases, exceptions are explicitly associated with an HTTP status code ([example](https://github.com/DataBiosphere/terra-workspace-manager/blob/4a33150f2143d163e89bfdda839e4ebf04f09b03/src/main/java/bio/terra/workspace/common/exception/ErrorReportException.java#L11)) and used to auto-generate an appropriate response ([example](https://github.com/DataBiosphere/terra-workspace-manager/blob/4a33150f2143d163e89bfdda839e4ebf04f09b03/src/main/java/bio/terra/workspace/app/controller/GlobalExceptionHandler.java#L33)).
+
+### Final
+
+*   **Classes** – unless a class has been designed and documented for inheritance, it _should_ be marked as final.
+*   **Function parameters **– _prefer_ not marking as final, due to the noisiness of the “final” keyword. Use where necessary for occasional semantic emphasis.
+*   **Instance variables** – _always_ mark as final if it is never reassigned.
+*   **Local variables** – use of final is _optional_. Short and simple methods can help avoid inappropriate variable reassignment.
+*   **Static variables** – almost _always_ mark as final.
+
+### Immutability
+
+Quoting from Effective Java: "immutable classes are easier to design, implement, and use than mutable classes. They are less prone to error and are more secure." See [recommendations from Effective Java](https://github.com/HugoMatilla/Effective-JAVA-Summary#15-minimize-mutability) on how to minimize mutability. [Immutables](https://immutables.github.io/) is a nice library for immutable data transfer objects, as is [Guava](https://github.com/google/guava/wiki/ImmutableCollectionsExplained) for immutable collections.
+
+### Java 8 features
+
+Use Java 8 features where helpful. See [https://leanpub.com/whatsnewinjava8/read](https://leanpub.com/whatsnewinjava8/read) and [https://www.oracle.com/java/technologies/javase/8-whats-new.html](https://www.oracle.com/java/technologies/javase/8-whats-new.html) for more details.
+
+Here is a quick run-down of Java 8 topics:
+
+*   **java.time.* – **the JavaTime library _should_ be used for date-time manipulation. Duration is especially common. For non-core extensions, consider the [https://www.threeten.org/threeten-extra/](https://www.threeten.org/threeten-extra/) library.
```

Comment:
I'm not convinced I need another time library. I know Joda used to be the cool one, until Java 8 time made it obsolete.

Reply:
Ack. The exhortation is merely to _consider_ the option as needed. It's intended as a pointer to refer to before someone starts writing similar functionality on their own.